neqo-transport/src/connection.rs

```diff
@@ -634,88 +633,106 @@ impl Connection {
+            self.process_handshake(hdr, &d)?;
+            self.process_migrations(&d)?;
+        }
+        Ok(())
+    }
 
-            if matches!(self.state, State::WaitInitial) {
-                if self.role == Role::Server {
-                    assert!(matches!(hdr.tipe, PacketType::Initial(..)));
-                    // A server needs to accept the client's selected CID during the handshake.
-                    self.valid_cids.push(hdr.dcid.clone());
-                    // Install a path.
-                    assert!(self.paths.is_none());
-                    let mut p = Path::new(&d, hdr.scid.unwrap());
-                    p.local_cids.push(ConnectionId::generate(CID_LENGTH));
-                    self.paths = Some(p);
-
-                    // SecretAgentPreinfo::early_data() always returns false for a server,
-                    // but a non-zero maximum tells us if we are accepting 0-RTT.
-                    self.zero_rtt_state = if self.crypto.tls.preinfo()?.max_early_data() > 0 {
-                        ZeroRttState::Accepted
-                    } else {
-                        ZeroRttState::Rejected
-                    };
-                } else {
-                    qdebug!([self] "Changing to use Server CID={}", hdr.scid.as_ref().unwrap());
-                    let p = self
-                        .paths
-                        .iter_mut()
-                        .find(|p| p.received_on(&d))
-                        .expect("should have a path for sending Initial");
-                    p.remote_cid = hdr.scid.unwrap();
+    fn decrypt_body(&mut self, mut hdr: &mut PacketHdr, slc: &[u8]) -> Option<Vec<u8>> {
+        // Decryption failure, or not having keys is not fatal.
+        // If the state isn't available, or we can't decrypt the packet, drop
+        // the rest of the datagram on the floor, but don't generate an error.
+        let largest_acknowledged = self
+            .loss_recovery
+            .largest_acknowledged(PNSpace::from(hdr.epoch));
+        let res = match self.crypto.obtain_crypto_state(self.role, hdr.epoch) {
+            Ok(cs) => match cs.rx.as_ref() {
+                Some(rx) => {
+                    let pn_decoder = PacketNumberDecoder::new(largest_acknowledged);
+                    decrypt_packet(rx, pn_decoder, &mut hdr, slc)
                 }
-                self.set_state(State::Handshaking);
-            }
+                _ => Err(Error::KeysNotFound),
+            },
+            Err(e) => Err(e),
+        };
+        res.ok()
+    }
 
-            if !self.paths.iter().any(|p| p.received_on(&d)) {
-                // Right now, we don't support any form of migration.
-                // So generate an error if a packet is received on a new path.
-                return Err(Error::InvalidMigration);
-            }
+    /// Ok(true) if the packet is a duplicate
+    fn process_ack(&mut self, hdr: &PacketHdr, body: Vec<u8>, now: Instant) -> Res<bool> {
```

Comment:
Based on the commentary, `is_duplicate()` might be a better name for this one, but then I see that we actually process the packet in here.  I think that you might want to find a new name for `process_input` and then rename this to `process_input`.

Reply:
the input/output names seem pretty widely used, I wouldn't want to break it too much (unless it really makes sense).